Add `GetBoolXXX()` functions.

jmalloc opened the pull request:
Partially addresses #33 

This PR adds several new helper functions for reading configuration values from a `config.Bucket` and parsing them as a boolean.

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+// "false", "no", or "off", it returns false. These string comparisons are
+// case-insensitive. It returns an error if any other value is used.
+//
+// If they key is not defined, it returns true.
```
```suggestion
// If the key is not defined, it returns true.
```

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+	"strings"
+)
+
+// GetBoolT returns the value associated with the given key as a boolean.
```
After reading this doc block I was still unsure what the `T` was in the name or why it was returning `true` by default when the env var did not exist. I wonder if this should mention something about falling back to the default of true? If this is a common thing in Go libraries then feel free to ignore my this comment.

jmalloc replied:
Nah, I think you're right - these docs could be improved.

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+// "false", "no", or "off", it returns false. These string comparisons are
+// case-insensitive. It returns an error if any other value is used.
+//
+// If they key is not defined, it returns false.
```
```suggestion
// If the key is not defined, it returns false.
```

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+// "false", "no", or "off", it returns false. These string comparisons are
+// case-insensitive. It returns an error if any other value is used.
+//
+// If they key is not defined, it returns v.
```
```suggestion
// If the key is not defined, it returns v.
```

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	switch strings.ToLower(s) {
+	case "true", "yes", "on":
```
Should these include `"1"`, and the falsy ones below include `"0"`?

danilvpetrov replied:
I was about to write my comment on this but then i saw @koden-km's comment. I feel like the boolean value extraction should deal with `1` and `0` as well. For instance `DEBUG_LOGGING=1` looks quite idiomatic.

jmalloc replied:
I had them in there originally then removed them, just to be more explicit (I also had `t`, `f`, `y` and `n`).

My logic is that if you're using this you're generally going to be writing new software so it doesn't need to be compatible with existing environment variables. But, I could be convinced to include them again.

koden-km replied:
I thought `1` and `0` were the most common for boolean flags. I don't think I've ever seen `t`, `f`, `y` or `n` used before. so I wouldn't worry about those. The `true`/`false`, `on`/`off`, `yes`/`no` are clearer for everyone though and you wouldn't mistake those for integer values.

danilvpetrov replied:
I tend to agree, of all one-character values `0` and `1` definitely deserve staying.

jmalloc replied:
I'm not sure any of us agree :P  It sounds like Kev is arguing to allow `1` and `0`, and Danil is saying to allow all of the single-character variants? Do I have that right?


> I thought `1` and `0` were the most common for boolean flags

This *is* definitely common, but I can't actually think for a good argument why. Perhaps it's because in bash, et al, where you deal directly with strings you can compare it to `0` to check for truthiness, but we have a real `bool` type here in Go.

koden-km replied:
If its all for new software and the docs are clear, then i'm fine with the full word booleans and ignoring the single letter and `0` and `1`.

jmalloc replied:
If nothing else we can always add support for these "shortcut" values later if we find we really want them, but it would be a BC break to add them now then take them away.

danilvpetrov replied:
My opinion is to leave `1` and `0` pair in, but to disallow any other one-character values; though, I am flexible with retaining only `true/false`, `on/off`, and `yes/no` thus adding clarity to the new software that will include `dodeca`.

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+// "false", "no", or "off", it returns false. These string comparisons are
+// case-insensitive. It panics if any other value is used.
+//
+// If they key is not defined, it returns true.
```
```suggestion
// If the key is not defined, it returns true.
```

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+// "false", "no", or "off", it returns false. These string comparisons are
+// case-insensitive. It panics if any other value is used.
+//
+// If they key is not defined, it returns false.
```
```suggestion
// If the key is not defined, it returns false.
```

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,97 @@
+// "false", "no", or "off", it returns false. These string comparisons are
+// case-insensitive. It panics if any other value is used.
+//
+// If they key is not defined, it returns v.
```
```suggestion
// If the key is not defined, it returns v.
```

danilvpetrov reviewed (commented):
I wrote a comment next to @koden-km's to consider adding `1`-`0` pair for the boolean parsing.

jmalloc commented:
I'll rejig the docs a bit and add some better tests.

jmalloc commented:
I've added two additional functions `GetBool()` and `MustGetBool()`, and added more thorough Ginkgo-based tests. I've kept basic example test for each function.

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,149 @@
+	return GetBoolDefault(b, k, false)
+}
+
+// GetBoolDefault returns boolean representation of the value associated with k,
```
```suggestion
// GetBoolDefault returns the boolean representation of the value associated with k,
```

koden-km commented on config/bool.go:
```diff
@@ -0,0 +1,149 @@
+	return MustGetBoolDefault(b, k, false)
+}
+
+// MustGetBoolDefault returns boolean representation of the value associated
```
```suggestion
// MustGetBoolDefault returns the boolean representation of the value associated
```

koden-km reviewed (changes requested):
The docs are much better now. Just two docs lines feel like they are missing a "the" that the other doc blocks have.
